src/lib/fetch-release-data.ts

```diff
@@ -0,0 +1,96 @@
+import { GetStaticPropsResult } from 'next'
+import semverSort from 'semver/functions/rsort'
+import { Products as HashiCorpProduct } from '@hashicorp/platform-product-meta'
+import { Product } from 'types/products'
+import { makeFetchWithRetry } from './fetch-with-retry'
+
+export type OperatingSystem =
+  | 'darwin'
+  | 'freebsd'
+  | 'openbsd'
+  | 'netbsd'
+  | 'archlinux'
+  | 'linux'
+  | 'windows'
+
+export interface ReleaseVersion {
+  name: HashiCorpProduct
+  version: string
+  shasums: string
+  shasums_signature: string
+  builds: {
+    name: HashiCorpProduct
+    version: string
+    os: OperatingSystem
+    arch: string
+    filename: string
+    url: string
+  }[]
+}
+export interface ReleasesAPIResponse {
+  name: HashiCorpProduct
+  versions: {
+    [versionNumber: string]: ReleaseVersion
+  }
+}
+
+/**
+ * There is a bit of a race condition with product releases and the metadata for the latest release
+ * propagating to releases.hashicorp.com. Often all it takes is a re-deploy of the website for it to work,
+ * so we're introducing a retry when fetching the release data in hopes that we can avoid manual intervention.
+ */
+const fetchWithRetry = makeFetchWithRetry(fetch, { retries: 3, delay: 1000 })
+
+// exclude pre-releases and such
+const validSemverRegex = /^\d+\.\d+\.\d+$/
+
+function getLatestVersionFromVersions(versions: string[]): string {
+  // exclude pre-releases and/or versions with tags or extra metadata
+  const validVersions = versions.filter((version) =>
+    version.match(validSemverRegex)
+  )
+
+  // using the reverse sort here to get the latest version first
+  const [latestVersion] = semverSort(validVersions)
+
+  return latestVersion
+}
+
+export function generateStaticProps(
+  product: Product
+): Promise<GetStaticPropsResult<{ releases: ReleasesAPIResponse }>> {
+  const productSlug = product.slug
+
+  return fetchWithRetry(
+    `https://releases.hashicorp.com/${productSlug}/index.json`,
+    {
+      headers: {
+        'Cache-Control': 'no-cache',
+      },
+    }
+  )
+    .then((res) => res.json())
+    .then((result) => {
+      const latestVersion = getLatestVersionFromVersions(
+        Object.keys(result.versions)
+      )
+
+      return {
+        // 5 minutes
+        revalidate: 300,
+        props: {
+          releases: result,
+          product,
+          latestVersion,
+        },
+      }
+    })
+    .catch(() => {
+      throw new Error(
+        `--------------------------------------------------------
+        Unable to resolve release data on releases.hashicorp.com from link
+        <https://releases.hashicorp.com/${productSlug}/index.json>.
+        ----------------------------------------------------------`
+      )
+    })
+}
```

Comment:
The change to passing the full product object here diverges the implementation and is something we'll need to address when merging the assembly branch back into main, either updating all of the usages or supporting either a slug or a product.

What do you think about supporting either a slug OR a product object?

Reply:
We actually _need_ a full `product` object returned from this function so that the `CurrentProductContext` will be initialized correctly in `_app.tsx`, so I think this will only work if an object is passed as `product`. Do we add a TODO somewhere or an Asana task for now to make sure this divergence is addressed correctly when merging the UI assembly branch?

Reply:
The io sites don't need/use the `CurrentProductContext` at the moment, so supporting the full product object or a slug here would let us have both implementations.

Reply:
Will we have both cases running at the same time? I don't think we've decided when we're going to merge the assembly branch yet, so I'm not sure on that.

Reply:
The io code is structured so that it can live alongside the UI work on the same branch, but yeah I guess we haven't made a definitive call on it yet. It depends on how we're going to be rolling out the refreshed UI I suppose. It's a possibility.

It seems like if we can support both a string or the product object we can avoid headache later. Or do you see downsides to supporting both?

Reply:
I think long term we shouldn't support both because it makes the behavior more ambiguous, but to avoid headaches in the short term we can do both. I'll make the update and also add a TODO comment so we don't lose track of the long term goal.

Reply:
Updated in https://github.com/hashicorp/dev-portal/pull/117/commits/d47455687d53e00c30b8a6ec621372c72512c45e!